users/rLapz/server.c

```diff
@@ -1,219 +1,494 @@
+ * NOTE: 
+ *       true    : 1,   false  : 0    [ boolean      ]
+ *       success : >=0, failed : <0   [ return value ]
  */
 
-#include <fcntl.h>
+#include <endian.h>
+#include <errno.h>
+#include <inttypes.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <poll.h>
 #include <unistd.h>
 
 #include <arpa/inet.h>
+#include <netinet/in.h>
 #include <sys/types.h>
 #include <sys/socket.h>
+#include <netdb.h>
+
+#include "ftransfer.h"
+
+
+
+struct client { /* <-----------------------------. */
+	uint16_t port;                       /*  | */
+	int      fd;                         /*  | */
+	struct   sockaddr_storage addr;      /*  | */
+	char     addr_str[INET6_ADDRSTRLEN]; /*  | */
+};                                           /*  | */
+                                             /*  | */
+struct server {                              /*  | */
+	int listener;                        /*  | */
+                                             /*  | */
+	struct sockaddr_storage *addr;       /*  | */
+	struct client *client; /* >--------------' */
+	struct {
+		int fd_count, fd_size, poll_count;
+		struct pollfd *pfds;
+	} fds;
+};
 
-/* local include */
-#include "data_structure.h"
 
-/* macros */
-#define TRUE	0
-#define FALSE	1
 
 /* function declarations */
-static void file_handler(int *client_fd);
-static int inet_handler(const char *addr, uint16_t port);
+static int         get_listener(const char *addr, const char *port);
+static int         add_to_pfds(struct server *s, const int new_fd);
+static void        server_poll(struct server *s);
+static void        delete_from_pfds(struct server *s, const int index);
+static const char *get_addr_str(char *dest, const struct sockaddr *s);
+static uint16_t    get_port(const struct sockaddr *s);
+static int         recv_all(char *buffer, const size_t buffer_size,
+			size_t *recv_size, const int client_fd);
+static int         get_file_prop(const struct client *c, packet_t *prop);
+static void        file_io(const struct server *s, const int index);
+static void        clean_up(struct server *s);
 
-/* global variable declarations */
-/* anyway, there is no global variable here, yay! */
 
+/* global variables */
+extern int is_interrupted;
 
-int
-main(int argc, char *argv[])
+
+/* function implementations */
+static int
+get_listener(const char *addr, const char *port)
 {
-	if (argc < 3 || argc > 3) {
-		printf("Usage:\n\t%s [address] [port]\n", argv[0]);
-		return EXIT_FAILURE;
+	int ret, rv, yes = 1;
+	struct addrinfo hints = {0}, *ai, *p;
+
+	hints.ai_family   = AF_UNSPEC;   /* IPv4 or IPv6 */
+	hints.ai_socktype = SOCK_STREAM; /* TCP */
+
+	if ((rv = getaddrinfo(addr, port, &hints, &ai)) != 0) {
+		fprintf(stderr, "server: get_listener(): getaddrinfo: %s\n",
+			gai_strerror(rv)
+		);
+
+		return -1;
 	}
 
-	return inet_handler(argv[1], (uint16_t)atoi(argv[2]));
+	for (p = ai; p != NULL; p = p->ai_next) {
+		ret = socket(p->ai_family, p->ai_socktype, p->ai_protocol);
```

Comment:
> getaddrinfo() does not support this option (SOCK_NONBLOCK), so it will fail.

Sorry, I was wrong. So here?
It should be where your TCP server socket to be bound.
```suggestion
		ret = socket(p->ai_family, p->ai_socktype | SOCK_NONBLOCK, p->ai_protocol);
```

Reply:
> yeah, it works, but I prefer to use `fcntl` instead.
> 
> `fcntl(ret, F_SETFL, O_NONBLOCK);`

You would need 2 calls to retain the old flags if there are any.
```c
int set_nonblock(int fd)
{
	int old_flags;
	old_flags = fcntl(ret, F_GETFL, 0);
	if (old_flags < 0) {
		perror("fcntl");
		return -1;
	}

	old_flags |= O_NONBLOCK;
	if (fcntl(ret, F_SETFL, old_flags) < 0) {
		perror("fcntl");
		return -1;
	}
	return 0;
}
```